src/nanoemoji/svg.py

```diff
@@ -0,0 +1,189 @@
+from fontTools import ttLib
+from lxml import etree  # pytype: disable=import-error
+from nanoemoji.color_glyph import ColorGlyph, PaintedLayer
+from nanoemoji.disjoint_set import DisjointSet
+from picosvg.svg import to_element, SVG
+from picosvg import svg_meta
+from picosvg.svg_transform import Affine2D
+import regex
+from typing import MutableMapping, Sequence, Tuple
+
+
+def _create_doclist(svg: SVG) -> str:
+    return (
+        svg
+        # dumb sizing isn't useful
+        .remove_attributes(("width", "height"))
+        # Firefox likes to render blank if present
+        .remove_attributes(("enable-background",))
+    )
+
+
+def _ensure_has_id(el: etree.Element):
+    if "id" in el.attrib:
+        return
+    nth_child = 0
+    prev = el.getprevious()
+    while prev is not None:
+        nth_child += 1
+        prev = prev.getprevious()
+    el.attrib["id"] = f'{el.getparent().attrib["id"]}::{nth_child}'
+
+
+def _glyph_groups(color_glyphs: Sequence[ColorGlyph]) -> Tuple[Tuple[str, ...]]:
+    """Find glyphs that need to be kept together by union find."""
+    # glyphs by reuse_key
+    glyphs = {}
+    reuse_groups = DisjointSet()
+    for color_glyph in color_glyphs:
+        reuse_groups.make_set(color_glyph.glyph_name)
+        for painted_layer in color_glyph.as_painted_layers():
+            # TODO what attributes should go into this key for SVG
+            reuse_key = _inter_glyph_reuse_key(painted_layer)
+            if reuse_key not in glyphs:
+                glyphs[reuse_key] = color_glyph.glyph_name
+            else:
+                reuse_groups.union(color_glyph.glyph_name, glyphs[reuse_key])
+
+    return reuse_groups.sorted()
+
+
+def _add_unique_gradients(
+    id_updates: MutableMapping[str, str],
+    svg_defs: etree.Element,
+    color_glyph: ColorGlyph,
+):
+    for gradient in color_glyph.picosvg.xpath("//svg:defs/*"):
+        gradient = copy.deepcopy(gradient)
+        curr_id: str = gradient.attrib["id"]
+        new_id = f"{color_glyph.glyph_name}::{curr_id}"
+        del gradient.attrib["id"]
+        gradient_xml: str = etree.tostring(gradient)
+        if gradient_xml in id_updates:
+            id_updates[curr_id] = id_updates[gradient_xml]
+        else:
+            gradient.attrib["id"] = new_id
+            id_updates[curr_id] = new_id
+            id_updates[gradient_xml] = new_id
+            svg_defs.append(gradient)
+
+
+def _inter_glyph_reuse_key(painted_layer: PaintedLayer):
+    """Individual glyf entries, including composites, can be reused.
+
+    SVG reuses w/paint so paint is part of key."""
+
+    # TODO we could recycle shapes that differ only in paint, would just need to
+    # transfer the paint attributes onto the use element if they differ
+    return (painted_layer.paint, painted_layer.path.d, painted_layer.reuses)
+
+
+def _add_glyph(svg, color_glyph, id_updates):
+    # each glyph gets a group of its very own
+    svg_g = svg.append_to("/svg:svg", etree.Element("g"))
```

Comment:
perhaps we could do away with a "g" wrapper if it's the only glyph in the svg document

Reply:
I kind of like the consistency. It makes predicting what our compiler does very easy. You have to have the id somewhere so this is just costing you the actual `<g></g>`.